### ISSUE TITLE
Implementing rmsprop stochastic optimisation for neural nets

### ISSUE THREAD

**Issue by arasuarun** [shogun-toolbox/shogun#2967]
A brief description of rmsprop, which is used in speeding up mini batch learning: http://www.cs.toronto.edu/~tijmen/csc321/slides/lecture_slides_lec6.pdf 
(3rd from last slide)

The “maths” part has already been written in /lib/optimisation/RmsPropUpdater.cpp and the training algorithm has to be implemented in the Neural Networks package

**Comment by arasuarun** [comment]
The current plan I have is to add data members required for rmsProp to NeuralNetwork.cpp with default values such that RmsProp doesn't make any difference.
The parameters required for rmsprop are (i.e. creating an RmspropUpdate object) :
1. learning rate (already in NN.cpp)
2. epsilon
3. decay_factor

In this way, there will always be an RmsProp update object in the training function regardless of whether it is being used or not. Would this be okay or would another way be better?

**Comment by arasuarun** [comment]
From the suggestions of @yorkerlin it makes more sense to have a general implementation, instead of just RmsProp. Hence, we can use AdaDeltaUpdater, AdaGradUpdater, RmsPropUpdater, etc... without changing the code in neural nets.

@iglesias @lisitsyn @yorkerlin My understanding so far is that they all derive from DescendUpdaterWithCorrection. The optimisation methods that are already there (and to come) have the same interface and (I think) the same data members. 
Please correct me if I'm wrong. Is there an even more general implementation possible?

**Comment by iglesias** [comment]
It makes sense to me.

**Comment by arasuarun** [comment]
So, Descend Updater is the base class for all these optimisation methods, which all have the "update_variable" method with the same parameters.
I have a doubt about the implementation: how exactly would the user choose which optimisation method he wants? Is it like we have a DescendUpdater *object as a data member in neuralnetwork.cpp* and the exact derived class can be chosen by constructing that? 
@yorkerlin

**Comment by yorkerlin** [comment]
From user' view,  users only need to pass a string (enum class) to use
existing optimizers.
For implementation, we may use factory pattern or strategy pattern.
Any other ideas?

On Thursday, January 21, 2016, Arasu Arun <notifications@github.com> wrote:

> So, Descend Updater is the base class for all these optimisation methods,
> which all have the "update_variable" method with the same parameters.
> I have a doubt about the implementation: how exactly would the user choose
> which optimisation method he wants? Is it like we have a DescendUpdater *object
> as a data member in neuralnetwork.cpp* and the exact derived class can be
> chosen by constructing that?
> @yorkerlin <https://github.com/yorkerlin>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/shogun-toolbox/shogun/issues/2967#issuecomment-173794249>
> .
>


-- 
best,

wu lin

**Comment by arasuarun** [comment]
@yorkerlin Sorry for the late reply. I'll check the patterns you mentioned and get on it right away. Thank you. :)

**Comment by arasuarun** [comment]
I understand the strategy method implementation. 
One question: if we use enums, how will the user choose among them? 
By giving the **integer** or the **string**?
```
enum DESC_UPDATER = { GRAD_DESC, ADA_DELTA, ADA_GRAD, RMS_PROP };

net->SetDescendUpdater(3, ...); 
// or
net->SetDescendUpdater("RMS_PROP", ...);
```

**Comment by arasuarun** [comment]
I've started #2971 for this issue. I followed the strategy pattern for the implementation.
